README
NAME
    GraphViz::DBI - Graph database tables and relations

VERSION
    version 1.100860

SYNOPSIS
      use GraphViz::DBI;
      my $dbh = DBI->connect('...', '...', '...');
      print GraphViz::DBI->new($dbh)->graph_tables->as_png;

DESCRIPTION
    This module constructs a graph for a database showing tables and
    connecting them if they are related. While or after constructing the
    object, pass an open database handle, then call "graph_tables" to
    determine database metadata and construct a GraphViz graph from the
    table and field information.

    The methods described here are defined by this class; all other method
    calls are passed to the underlying GraphViz object:

METHODS
  new([$dbh])
    Constructs the object; also creates a GraphViz object. The constructor
    accepts an optional open database handle.

  set_dbh($dbh)
    Sets the database handle.

  get_dbh()
    Returns the database handle.

  get_tables()
    Returns the tables defined in the database.

  is_table($table)
    Checks the database metadata whether the argument is a valid table name.

  is_foreign_key($table, $field)
    Determines whether the field belonging to the table is a foreign key
    into some other table. If so, it is expected to return the name of that
    table. If not, it is expected to return a false value.

    For example, if there is a table called "product" and another table
    contains a field called "product_id", then to indicate that this field
    is a foreign key into the product table, the method returns "product".
    This is the logic implemented in this class. You can override this
    method in a subclass to suit your needs.

  graph_tables()
    This method goes through all tables and fields and calls appropriate
    methods to determine which tables and which dependencies exist, then
    hand the results over to GraphViz. It returns the GraphViz object.

FEATURE IDEAS
    *   Test with various database drivers to see whether they support the
        metadata interface.

    *   Make each table a vertical port with dependencies using those ports.

    *   Provide the possibility to name edges to specify the type of
        relationship ('has-a', 'is-a', etc.).

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=GraphViz-DBI>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/GraphViz-DBI/>.

    The development version lives at
    <http://github.com/hanekomu/GraphViz-DBI/>. Instead of sending patches,
    please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2001 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension GraphViz-DBI

1.100860  2010-03-27 23:19:42 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.04  Thu Jul 25 15:27:43 CEST 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - Changed dist style and Changes back to standard. People didn't like it -
      the nail that sticks out gets hammered down.
    - Added standard test files; this will also help with CPANTS.

0.03  Thu, 29 May 2008 12:05:22 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - Converted Changes file to YAML style
    - .shipit: added Twitter step
    - Makefile.PL: added auto_install() and process_templates()
    - lib/*: converted to template
    - incorporated changes from Aaron Travena's version 0.02 from CPAN - had
      forgotten about that version...
    - added t/perlcriticrc
    - updated MANIFEST
    - .shipit: fixed svk.tagpattern
    - tags: NEWFEATURE

0.02  Sat, 29 Mar 2003 17:40:05 +0100 (Aaron Trevena <aaron.trevena@droogs.org>)
    - Updates

0.01  Mon, 05 Mar 2001 21:11:04 +0100 (Marcel Gruenauer <marcel@cpan.org>)
    - original version


eg/dbigraph.pl
#!/usr/bin/perl
use warnings;
use strict;
use Getopt::Long;
use Pod::Usage;
use DBI;
use GraphViz::DBI;
my %opts = (
    help    => 0,
    man     => 0,
    verbose => 0,
    dbd     => '',
    dbname  => '',
    dsn     => '',
    user    => '',
    pass    => '',
    as      => '',
);
GetOptions(
    \%opts, qw(
      help
      man
      verbose
      dbd=s
      dbname=s
      dsn=s
      user=s
      pass=s
      as=s
      )
) || pod2usage(2);
pod2usage(1) if $opts{help};
pod2usage(-exitstatus => 0, -verbose => 2) if $opts{man};
$opts{dsn} = "dbi:$opts{dbd}:dbname=$opts{dbname}"
  if $opts{dbd} && $opts{dbname};
pod2usage(1) unless $opts{dsn};
$opts{as} ||= 'png';
my $dbh = DBI->connect(@opts{qw/dsn user pass/});
my $as  = "as_$opts{as}";
print GraphViz::DBI->new($dbh)->graph_tables->$as;
$dbh->disconnect;
__END__

=head1 NAME

dbigraph.pl - graph database tables and their relations

=head1 SYNOPSIS

dbigraph.pl --dbd=Pg --dbname=mydb --user=marcel | xv -

dbigraph.pl --dsn='dbi:Pg:dbname=mydb' --user=marcel --as=png >mydb.png

=head1 DESCRIPTION

This program constructs a GraphViz graph for a database showing tables
and connecting them if they are related.

=head1 OPTIONS

This section describes the supported command line options. Minimum
matching is supported.

=over 4

=item B<--dbd>

Database drive to use, e.g. 'Pg'. Only if given together with the
C<--dbname> option will this be used for the DSN.

=item B<--dbname>

Database name to use. Only if given together with the C<--dbd> option
will this be used for the DSN.

=item B<--dsn>

DSN to use, e.g. 'dbi:Pg:mydb'. Only used if C<--dbd> and C<--dbname>
aren't given.

=item B<--user>

Username to use for connecting to the database.

=item B<--pass>

Password to use for connecting to the database.

=item B<--as>

Output format, e.g. 'gif', 'png', etc. Defaults to 'png' if not given.

=item B<--help>

Print a brief help message and exit.

=item B<--man>

Print the manual page and exit.

=item B<--verbose>

Print information messages as we go along.

=back

=head1 BUGS

Some. Possibly. I haven't fully tested it.

=head1 AUTHOR

Marcel GrE<uuml>nauer E<lt>marcel@codewerk.comE<gt>

=head1 COPYRIGHT

Copyright 2001-2009 Marcel GrE<uuml>nauer. All rights reserved.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

GraphViz(3pm), GraphViz::DBI(3pm).

=cut


lib/GraphViz/DBI.pm
use 5.008;
use strict;
use warnings;

package GraphViz::DBI;
our $VERSION = '1.100860';
# ABSTRACT: Graph database tables and relations
use Carp;
use GraphViz;
our $AUTOLOAD;

sub new {
    my $this  = shift;
    my $class = ref($this) || $this;
    my $self  = {};
    bless $self, $class;
    $self->_init(@_);
    return $self;
}

sub _init {
    my $self = shift;
    $self->set_dbh(+shift) if @_;
    $self->{g} = GraphViz->new();
}

sub set_dbh {
    my ($self, $dbh) = @_;
    $self->{dbh} = $dbh;
    return $self;
}

sub get_dbh {
    my $self = shift;
    return $self->{dbh};
}

sub get_tables {
    my $self = shift;
    $self->{tables} ||= [ $self->get_dbh->tables ];
    return @{ $self->{tables} };
}

sub is_table {
    my ($self, $table) = @_;
    $self->{is_table} ||= { map { $_ => 1 } $self->get_tables };
    return $self->{is_table}{$table};
}

sub is_foreign_key {

    # if the field name is of the form "<table>_id" and
    # "<table>" is an actual table in the database, treat
    # this as a foreign key.
    # This is my convention; override it to suit your needs.
    my ($self, $table, $field) = @_;
    return if $field =~ /$table[_-]id/i;
    return unless $field =~ /^(.*)[_-]id$/i;
    my $candidate = $1;
    return unless $self->is_table($candidate);
    return $candidate;
}

sub graph_tables {
    my $self = shift;
    my %table = map { $_ => 1 } $self->get_tables;
    for my $table ($self->get_tables) {
        my $sth = $self->get_dbh->prepare("select * from $table where 1 = 0");
        $sth->execute;
        my @fields = @{ $sth->{NAME} };
        $sth->finish;
        my $label = "{$table|";
        for my $field (@fields) {
            $label .= $field . '\l';
            if (my $dep = $self->is_foreign_key($table, $field)) {
                $self->{g}->add_edge({ from => $table, to => $dep });
            }
        }
        $self->{g}->add_node(
            {   name  => $table,
                shape => 'record',
                label => "$label}",
            }
        );
    }
    $self->{g};
}

sub AUTOLOAD {
    my $self = shift;
    my $type = ref($self) or croak "$self is not an object";
    (my $name = $AUTOLOAD) =~ s/.*:://;
    return if $name =~ /DESTROY/;

    # hm, maybe GraphViz knows what to do with it...
    $self->{g}->$name(@_);
}
1;


__END__
=pod

=head1 NAME

GraphViz::DBI - Graph database tables and relations

=head1 VERSION

version 1.100860

=head1 SYNOPSIS

  use GraphViz::DBI;
  my $dbh = DBI->connect('...', '...', '...');
  print GraphViz::DBI->new($dbh)->graph_tables->as_png;

=head1 DESCRIPTION

This module constructs a graph for a database showing tables and
connecting them if they are related. While or after constructing the
object, pass an open database handle, then call C<graph_tables> to
determine database metadata and construct a GraphViz graph from the
table and field information.

The methods described here are defined by this class; all other method calls
are passed to the underlying GraphViz object:

=head1 METHODS

=head2 new([$dbh])

Constructs the object; also creates a GraphViz object. The constructor
accepts an optional open database handle.

=head2 set_dbh($dbh)

Sets the database handle.

=head2 get_dbh()

Returns the database handle.

=head2 get_tables()

Returns the tables defined in the database.

=head2 is_table($table)

Checks the database metadata whether the argument is a valid table name.

=head2 is_foreign_key($table, $field)

Determines whether the field belonging to the table is a foreign key
into some other table. If so, it is expected to return the name of that
table. If not, it is expected to return a false value.

For example, if there is a table called "product" and another table
contains a field called "product_id", then to indicate that this field is
a foreign key into the product table, the method returns "product". This
is the logic implemented in this class. You can override this method in
a subclass to suit your needs.

=head2 graph_tables()

This method goes through all tables and fields and calls appropriate
methods to determine which tables and which dependencies exist, then
hand the results over to GraphViz. It returns the GraphViz object.

=head1 FEATURE IDEAS

=over 4

=item *

Test with various database drivers to see whether they support the
metadata interface.

=item *

Make each table a vertical port with dependencies using those ports.

=item *

Provide the possibility to name edges to specify the type of relationship
('has-a', 'is-a', etc.).

=back

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=GraphViz-DBI>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/GraphViz-DBI/>.

The development version lives at
L<http://github.com/hanekomu/GraphViz-DBI/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2001 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'GraphViz::DBI',
                       'DISTNAME' => 'GraphViz-DBI',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Graph database tables and relations',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100860',
                       'PREREQ_PM' => {
                                        'Getopt::Long' => '0',
                                        'Scalar::Util' => '0',
                                        'English' => '0',
                                        'Pod::Usage' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'File::Find' => '0',
                                        'GraphViz' => '0',
                                        'File::Temp' => '0',
                                        'DBI' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



